File: test/technology/inml/test_inml_layout.cpp
/*
 * Copyright (c) 2018 - 2023 Marcel Walter
 * Copyright (c) 2023 - present Chair for Design Automation, Technical University of Munich
 * All rights reserved.
 *
 * SPDX-License-Identifier: MIT
 *
 * Licensed under the MIT License
 */

/**
 * @file
 * @brief Tests for `fiction/technology/inml/layout.hpp`.
 * @author Marcel Walter (marcelwa)
 */

#include <catch2/catch_test_macros.hpp>

#include <fiction/layouts/clocking_scheme.hpp>
#include <fiction/technology/inml/layout.hpp>

using namespace fiction;
using namespace fiction::layouts;
using namespace fiction::inml;

TEST_CASE("iNML layouts hold planar, clocked magnets", "[inml-layout]")
{
    layout lyt{{7, 7, 1}, clocking::twoddwave(clocking::num_clks::THREE), "inverter", 4, 4};

    CHECK(lyt.z() == 0);
    CHECK(lyt.num_clocks() == 3);
    CHECK(lyt.get_tile_size_x() == 4);

    lyt.assign_cell_type({0, 1}, magnet_type::INPUT);
    lyt.assign_cell_type({1, 1}, magnet_type::INVERTER_MAGNET);
    lyt.assign_cell_type({5, 1}, magnet_type::OUTPUT);
    lyt.assign_cell_name({0, 1}, "a");

    CHECK(lyt.num_cells() == 3);
    CHECK(lyt.get_cell_type({1, 1}) == magnet_type::INVERTER_MAGNET);
    CHECK(lyt.get_cell_name({0, 1}) == "a");
    CHECK(lyt.get_clock_number({1, 1}) == 0);
    CHECK(lyt.get_clock_number({5, 1}) == 1);

    auto copy = lyt;
    CHECK(copy == lyt);
    copy.assign_cell_type({1, 1}, magnet_type::NORMAL);
    CHECK(lyt.get_cell_type({1, 1}) == magnet_type::INVERTER_MAGNET);
}

File: test/python/inml/test_inml_layout.py
# Copyright (c) 2018 - 2023 Marcel Walter
# Copyright (c) 2023 - present Chair for Design Automation, Technical University of Munich
# All rights reserved.
#
# SPDX-License-Identifier: MIT
#
# Licensed under the MIT License

from __future__ import annotations

from mnt.pyfiction.inml import inml_layout, inml_magnet_type
from mnt.pyfiction.layouts.coords import offset_coordinate


def test_magnet_types() -> None:
    assert str(inml_magnet_type.INVERTER_MAGNET) == "inml_magnet_type.INVERTER_MAGNET"
    assert str(inml_magnet_type.FANOUT_COUPLER_MAGNET) == "inml_magnet_type.FANOUT_COUPLER_MAGNET"


def test_planar_clocked_layout() -> None:
    layout = inml_layout(offset_coordinate(7, 7, 1), "2DDWave", "inverter", 4, 4)

    assert layout.z() == 0
    assert layout.get_tile_size_x() == 4

    layout.assign_cell_type(offset_coordinate(0, 1), inml_magnet_type.INPUT)
    layout.assign_cell_type(offset_coordinate(1, 1), inml_magnet_type.INVERTER_MAGNET)
    layout.assign_cell_type(offset_coordinate(5, 1), inml_magnet_type.OUTPUT)

    assert layout.num_cells() == 3
    assert layout.get_cell_type(offset_coordinate(1, 1)) == inml_magnet_type.INVERTER_MAGNET
    assert layout.get_clock_number(offset_coordinate(1, 1)) == 0
    assert layout.get_clock_number(offset_coordinate(5, 1)) == 1
